[INTERNAL][STF] NetworkManipulator should work again

srossbach opened the pull request:
It now uses the transmitter instead of the connection to send packets
which caused issues because the content was not compressed although it
should.

srossbach commented:
@m273d15 could you please test this patch ? My STF is not working. And I have no clue what is causing the NPE.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/internal/XMPPTransmitter.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5135213&bid=16873289)

tobous commented:
> @m273d15 could you please test this patch ? My STF is not working. And I have no clue what is causing the NPE.

@srossbach See my response in #854 (https://github.com/saros-project/saros/issues/854#issuecomment-595792445).

Also, for future reference: You can also run the STF through our build infrastructure (e.g. if you don't want to run it yourself). To do so, simply name your branch `pr/stf/*` and push to the repo. An STF build will then also be started as a GitHub action. This build will also be displayed in the matching PR for the branch.

m273d15 commented:
My results of your branch:
The test still fails, but with the exception (in ALICE's eclipse instance):
```
TRACE 16:00:28,044 [BinaryChannel-dev0_carl_stf] (BinaryChannelConnection.java:308) processing opcode 0xFA [TFD]: id=0, chunks=1
TRACE 16:00:28,045 [BinaryChannel-dev0_carl_stf] (BinaryChannelConnection.java:346) processing opcode 0xFB [DATA]: id=0, DATA len=89 bytes
TRACE 16:00:28,045 [saros-DispatchContext] (NetworkManipulatorImpl.java:99) intercepting incoming packet from: dev0_carl_stf@saros-con.imp.fu-berlin.de/Saros
TRACE 16:00:28,045 [saros-DispatchContext] (XMPPReceiver.java:214) received binary XMPP extension: TransferDescription [elementName=ados, namespace=saros, recipient=dev0_alice_stf@saros-con.imp.fu-berlin.de/Saros, sender=dev0_carl_stf@saros-con.imp.fu-berlin.de/Saros, compress=true], size: 89, RX time: 0 ms [SOCKS5 (D)]
ERROR 16:00:28,046 [saros-DispatchContext] (ThreadUtils.java:36) Internal Error:
java.lang.NullPointerException
	at saros.session.internal.ActivitySequencer.receiveActivities(ActivitySequencer.java:572)
	at saros.session.internal.ActivitySequencer.access$0(ActivitySequencer.java:540)
	at saros.session.internal.ActivitySequencer$1.processPacket(ActivitySequencer.java:97)
	at saros.net.internal.XMPPReceiver.forwardPacket(XMPPReceiver.java:192)
	at saros.net.internal.XMPPReceiver.access$1(XMPPReceiver.java:181)
	at saros.net.internal.XMPPReceiver$6.run(XMPPReceiver.java:171)
	at saros.util.ThreadUtils$1.run(ThreadUtils.java:34)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
ERROR 16:00:28,046 [saros-DispatchContext] (ThreadUtils.java:37) Original caller:
StackTrace:
	at saros.util.ThreadUtils.wrapSafe(ThreadUtils.java:28)
	at saros.net.DispatchThreadContext.executeAsDispatch(DispatchThreadContext.java:27)
	at saros.net.internal.XMPPReceiver.receive(XMPPReceiver.java:163)
	at saros.net.internal.BinaryChannelConnection$ReceiverThread.run(BinaryChannelConnection.java:95)
```

srossbach commented:
I will try to get the STF working again and will post and update.

srossbach commented:
After testing it seems that XStream is unable to find specific classes. When calling XStream from Saros/E to STF code (interceptor) toXML produces the correct result, however when calling XStream from STF all activities inside the extension are simply not marshalled.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/internal/XMPPTransmitter.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Saros/saros/pullRequest?prid=5135213&bid=16873289)

srossbach commented:
> @srossbach See my response in #854 ([#854 (comment)](https://github.com/saros-project/saros/issues/854#issuecomment-595792445)).

I already posted another issues (bundle cache out of date) and how to fix it. Maybe it isn't a bad idea to actually list those issues in the documentation and how to solve them.

srossbach commented:
It depends if you want to store the already marshalled and compressed content or just the "plain" Smack packet. If you see the comment in the ActivitySequencer then the sequencer assumed that the packet is immediately marshalled.  I modified the Interceptor in such a way to make it easier to just store packets and removed the need to access the IConnection (which on the otherhand would bypass the whole ITransmitter) when resending the stored packets.

srossbach commented on core/src/saros/net/internal/XMPPTransmitter.java:
```diff
@@ -79,6 +79,13 @@ public void send(JID recipient, PacketExtension extension) throws IOException {
   public void send(String connectionID, JID recipient, PacketExtension extension)
       throws IOException {
 
+    boolean sendPacket = true;
```
See bug description. The content was flagged as compressed but it was not. Of you I have could modified it in such a way that the compression was actually done before the interceptor is called. After thinking about the problem I thought that just intercepting the whole packet is easier than: take that bunch of bytes along with a transfer description, deal with it.
